vllm/platforms/cuda.py

```diff
@@ -144,7 +109,118 @@ def is_full_nvlink(cls, physical_device_ids: List[int]) -> bool:
                         return False
         return True
+
+    @classmethod
+    @lru_cache(maxsize=8)
+    def _get_physical_device_capability(cls,
+                                        device_id: int = 0) -> Tuple[int, int]:
+        handle = pynvml.nvmlDeviceGetHandleByIndex(device_id)
+        return pynvml.nvmlDeviceGetCudaComputeCapability(handle)
+
+    @classmethod
+    @lru_cache(maxsize=8)
+    def _get_physical_device_name(cls, device_id: int = 0) -> str:
+        handle = pynvml.nvmlDeviceGetHandleByIndex(device_id)
+        return pynvml.nvmlDeviceGetName(handle)
+
+    @classmethod
+    @lru_cache(maxsize=8)
+    def _get_physical_device_total_memory(cls, device_id: int = 0) -> int:
+        handle = pynvml.nvmlDeviceGetHandleByIndex(device_id)
+        return int(pynvml.nvmlDeviceGetMemoryInfo(handle).total)
+
+    @classmethod
+    def log_warnings(cls):
+        device_ids: int = pynvml.nvmlDeviceGetCount()
+        if device_ids > 1:
+            device_names = [
+                cls._get_physical_device_name(i) for i in range(device_ids)
+            ]
+            if len(set(device_names)) > 1 and os.environ.get(
+                    "CUDA_DEVICE_ORDER") != "PCI_BUS_ID":
+                logger.warning(
+                    "Detected different devices in the system: \n%s\nPlease"
+                    " make sure to set `CUDA_DEVICE_ORDER=PCI_BUS_ID` to "
+                    "avoid unexpected behavior.", "\n".join(device_names))
+
+
+class NonNVMLContext(BaseContext):
+
+    @classmethod
+    def get_device_capability(cls, device_id: int = 0) -> Tuple[int, int]:
+        return torch.cuda.get_device_capability(device_id)
+
+    @classmethod
+    def get_device_name(cls, device_id: int = 0) -> str:
+        return torch.cuda.get_device_name(device_id)
+
+    @classmethod
+    def get_device_total_memory(cls, device_id: int = 0) -> int:
+        device_props = torch.cuda.get_device_properties(device_id)
+        return device_props.total_memory
+
+    @classmethod
+    def is_full_nvlink(cls, physical_device_ids: List[int]) -> bool:
+        logger.exception(
+            "NVLink detection not possible, as context support was"
+            " not found. Assuming no NVLink available.")
+        return False
+
+
+@contextmanager
+def get_context() -> Iterator[BaseContext]:
+    nvml_init_ok = False
+    try:
+        try:
+            pynvml.nvmlInit()
+            nvml_init_ok = True
+            yield NVMLContext()
+        except Exception:
+            # On Jetson, NVML is not supported.
+            yield NonNVMLContext()
+    finally:
+        if nvml_init_ok:
+            pynvml.nvmlShutdown()
+
+
+try:
+    from sphinx.ext.autodoc.mock import _MockModule
+
+    if not isinstance(pynvml, _MockModule):
+        with get_context() as context:
+            context.log_warnings()
+except ModuleNotFoundError:
+    with get_context() as context:
+        context.log_warnings()
+
+
+class CudaPlatform(Platform):
```

Comment:
we can have `NvmlCudaPlatform` and `NonNvmlCudaPlatform` inheriting from `Platform`, and in the end of this file, based on jetson or not, define a variable `CudaPlatform` to point to either `NvmlCudaPlatform` or `NonNvmlCudaPlatform` .